apache/jmeter
# 63130

## Description

Class name: HTTPArgument.java
Method name: HTTPArgument (constructor) 
Class name: HttpHC4Impl.java
Method name: createUrlEncodedFormEntity

Replaced java.net.URLDecoder class from org.apache.commons.codec.net.URLCodec class


## Motivation and Context

JMeter uses java.net.URLDecoder class for URL encoding.
Its drawback : This library can not encode Shift_JIS correctly because of different specific interpretation between Java and IE.
Because of browser specification, char code is zipped.
Browser encode Japanese text as following.
初期値大阪市 ⇒ %8F%89%8A%FA%92l%91%E5%8D%E3%8Es
初 = %8F%89
期 = %8A%FA
値 = %92l      ← % is lack
大 = %91%E5
阪 = %8D%E3
市 = %8Es     ← % is lack

JMeter should use org.apache.commons.codec.net.URLCodec class for URL encoding.


https://bz.apache.org/bugzilla/show_bug.cgi?id=63130 

## How Has This Been Tested?


1. I have created one sample application, which contain some pages having shift_jis encoding, and deployed the same in tomcat server.
2. Created the test plan in Jmeter tool and configure it so that it can receive all the incoming request from the application.
3. Start the Http(s) Test Script Recorder
4. Run the application, and post the request from the application.
5. Validate the result in response page.



Run the "ant test" command to be ensured that no existing tests are breaking because of these changes.
Run the "ant package-and-check" to check the coding guidelines.

Other Details: 
- OS: windows 8
- Java: 8
- Jmeter version: 5.0
- Hardware: PC 


## Screenshots (if appropriate):

## Types of changes

- Bug fix (non-breaking change which fixes an issue)


## Checklist:


- [x ] My code follows the [code style][style-guide] of this project.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines
